design/one-pager-package-image-config.md

```diff
@@ -0,0 +1,377 @@
+            url: https://fulcio.sigstore.dev
+            identities:
+              - issuer: https://token.actions.githubusercontent.com
+                subjectRegExp: https://github.com/acme-co/crossplane-packages/*
+          attestations:
+            - name: verify attestations
+              predicateType: spdxjson
+```
+
+The `ImageConfig` API `spec` has the following fields:
+
+- `matchImages`: A list of image prefixes to match. The package manager will
+  apply the settings defined in the `ImageConfig` object to the images that
+  match the prefixes. The longest prefix match will be selected in case of
+  multiple matches.
+- `registry`: Configuration for interacting with the registry hosting the images.
+  - `authentication`: Authentication settings for the registry.
+    - `pullSecretRef`: Reference to the Kubernetes Secret containing the
+      credentials to pull images from the registry. This secret must be of type
+      [`kubernetes.io/dockerconfigjson`].
+  - `tls`: TLS settings for the registry.
+    - `enabled`: Whether to enable TLS for the registry. Default is `true`.
+    - `insecure`: Whether to skip verifying the server's certificate. Default is
+      `false`.
+    - `caBundleConfigMapRef`: Reference to the ConfigMap containing the CA
+      certificate bundle to use for verifying the server's certificate.
+- `verification`: Configuration for verifying image signatures.
+  - `provider`: The provider to use for verifying image signatures. In the
+    beginning, only `cosign` will be supported.
+  - `cosign`: Configuration for verifying images using cosign.
+    - `authorities`: List of authorities to use for verifying images.
+      - `name`: The name of the authority.
+      - `keyless`: Keyless verification settings.
+        - `url`: The URL of the keyless authority.
+        - `identities`: List of identities to use for verification.
+          - `issuer`: The issuer of the identity.
+          - `subjectRegExp`: The regular expression to match the subject of the
+            identity.
+      - `attestations`: List of attestations to use for verification.
+        - `name`: The name of the attestation.
+        - `predicateType`: The type of the predicate to use for verification.
+
+The `ImageConfig` object will be a global object in the cluster, configuring
+the package manager behavior without being referenced by any other object. There
+could be multiple `ImageConfig` objects in the cluster, each defining settings
+for different sets of images.
+
+When a package image needs to be pulled, either as a primary package or as a
+dependency, the package manager will search for `ImageConfig` objects in the
+cluster that match the image. The package manager selects the best match based
+on the longest prefix match with the image. If there are multiple matches with
+the same prefix length (i.e., identical prefixes across `ImageConfig`s), the
+package manager will select one arbitrarily. Similarly, after the image is
```

Comment:
I'm personally not a fan of `arbitrary selection`, it usually leads to confusion for users and subtle bugs for us. Did you have something specific in mind? Sorting `ImageCatalogs` by `metadata.name` maybe?

Reply:
Random selection has its place, but I agree it's not here.

I think the most obvious, least surprising thing to do would be to return an error if we can't figure out what configuration to use.

The challenge is that we don't want to surprisingly or subtly switch what image config we're using for a particular package. If we go with the lexicographically earliest ImageConfig name, I can imagine a scenario where everything works fine for months but then suddenly breaks because someone added a new ImageConfig that wasn't supposed to affect a particular package, but now does because it has a matching prefix and an earlier name. Or maybe it doesn't break, but the new config silently works in a different way than it should - e.g. allowing packages that shouldn't be allowed.

Reply:
If we don't need to introduce regexps or globs as I explained [here](https://github.com/crossplane/crossplane/pull/5993#discussion_r1794839623), we could improve the experience around simple prefixes. I mean:

> If there are multiple matches with the same prefix length (i.e., identical prefixes across `ImageConfig`s), the package manager will select one arbitrarily

We could introduce a validating webhook which ensure there is no duplicate prefixes across ImageConfigs, which will remove the need to arbitrary selection. The only case I am not sure about is that one may want to have duplicate prefixes if they want to configure pullSecret with one image config but verification with another one.

Reply:
The validating webhook won't work if we add more complex matching rules that we can't easily sort later on.

Reply:
how does Kyverno handle multiple matches?

Reply:
> The validating webhook won't work if we add more complex matching rules

More complex matching rules complicates everything, this is why I want to avoid as long as we don't have to :)

Reply:
> We could introduce a validating webhook which ensure there is no duplicate prefixes across ImageConfigs, which will remove the need for arbitrary selection.

This SGTM. I'd also be happy with a controller error if/when multiple configs match at least to begin with, since I assume that'd be easier to build than a webhook. The webhook approach does sound like a better UX.

Reply:
Actually now that I think of it, couldn't most of the `ImageConfig` settings be AND-ed?

I think this is a similar question/observation to https://github.com/crossplane/crossplane/pull/5993#discussion_r1793085002, except that I'm thinking it would only happen when there are multiple prefixes with the same length - i.e. multiple `ImageConfig` resources with identical `prefix` entries.

For example consider the following two configs:

```yaml
apiVersion: pkg.crossplane.io/v1alpha1
kind: ImageConfig
metadata:
  name: acme-sigstore
spec:
  matchImages:
    - prefix: registry1.com/acme-co/configuration-foo
  verification:
    provider: Cosign
    cosign:
      authorities:
        - name: verify acme builds
          keyless:
            url: https://fulcio.sigstore.dev
            identities:
              - issuer: https://token.actions.githubusercontent.com
                subjectRegExp: https://github.com/acme-co/crossplane-packages/*
          attestations:
            - name: verify attestations
              predicateType: spdxjson
```

```yaml
apiVersion: pkg.crossplane.io/v1alpha1
kind: ImageConfig
metadata:
  name: acme-internal-fulcio
spec:
  matchImages:
    - prefix: registry1.com/acme-co/configuration-foo
  verification:
    provider: Cosign
    cosign:
      authorities:
        - name: verify acme builds
          keyless:
            url: https://fulcio.internal.acme.co  # Note different URL
            identities:
              - issuer: https://token.actions.githubusercontent.com
                subjectRegExp: https://github.com/acme-co/crossplane-packages/*
          attestations:
            - name: verify attestations
              predicateType: spdxjson
```

Would it potentially make sense for these two configs to stack? i.e. The image must be verified by both Fulcio instances?

I could imagine pull secrets being the same - by having two `ImageConfig` with the same prefix and different pull secrets you'd effectively be saying "try both". Similar to how we'll try both if there's an `ImageConfig` that matches a package, and the package also has its own `imagePullSecrets`.

It seems like the only thing that wouldn't be AND-able would be the `tls` settings. Perhaps we could only return an error if multiple image configs specified TLS configurations for the same prefix?

Reply:
> couldn't most of the ImageConfig settings be AND-ed?

We could, but I don't think it makes much sense in practice that is worth complicating the behavior and the code. 

> Would it potentially make sense for these two configs to stack? i.e. The image must be verified by both Fulcio instances?

`authorities` already is an array allowing to pass multiple, I cannot think of a reason for someone _intentionally_ doing something like that.

Overall, it feels cleaner to say, "we will select one image config based on the match and use it" than, "in some cases we will stack them but in other cases we will pick one", especially considering there is no known / real use case for that right now. If we realize there are strong reasons to do so along the road, we can improve that.

Reply:
Resolving in favor of https://github.com/crossplane/crossplane/pull/5993#issuecomment-2406859296